DataManagementSystem/Agent/RequestOperations/ReplicateAndRegister.py

```diff
@@ -239,262 +245,318 @@ def _addMetadataToFiles( self, toSchedule ):
         opFileToSchedule.ChecksumType = metadata[lfn]['ChecksumType']
       opFileToSchedule.Size = metadata[lfn]['Size']
 
-      filesToScheduleList.append( ( opFileToSchedule.toJSON()['Value'],
-                                    toSchedule[lfn][1],
-                                    toSchedule[lfn][2] ) )
-
-    return S_OK( filesToScheduleList )
-
+      filesToScheduleList.append((opFileToSchedule.toJSON()['Value'],
+                                  toSchedule[lfn][1],
+                                  toSchedule[lfn][2]))
 
+    return S_OK(filesToScheduleList)
 
-  def _filterReplicas( self, opFile ):
+  def _filterReplicas(self, opFile):
     """ filter out banned/invalid source SEs """
-    return filterReplicas( opFile, logger = self.log, dataManager = self.dm )
+    return filterReplicas(opFile, logger=self.log, dataManager=self.dm)
 
-  def ftsTransfer( self ):
+  def ftsTransfer(self):
     """ replicate and register using FTS """
 
-    self.log.info( "scheduling files in FTS..." )
+    self.log.info("scheduling files in FTS...")
 
     bannedTargets = self.checkSEsRSS()
     if not bannedTargets['OK']:
-      gMonitor.addMark( "FTSScheduleAtt" )
-      gMonitor.addMark( "FTSScheduleFail" )
+      gMonitor.addMark("FTSScheduleAtt")
+      gMonitor.addMark("FTSScheduleFail")
       return bannedTargets
 
     if bannedTargets['Value']:
-      return S_OK( "%s targets are banned for writing" % ",".join( bannedTargets['Value'] ) )
+      return S_OK("%s targets are banned for writing" % ",".join(bannedTargets['Value']))
 
     # Can continue now
-    self.log.verbose( "No targets banned for writing" )
+    self.log.verbose("No targets banned for writing")
 
     toSchedule = {}
 
+    delayExecution = 0
+    errors = defaultdict(int)
     for opFile in self.getWaitingFilesList():
       opFile.Error = ''
-      gMonitor.addMark( "FTSScheduleAtt" )
+      gMonitor.addMark("FTSScheduleAtt")
       # # check replicas
-      replicas = self._filterReplicas( opFile )
+      replicas = self._filterReplicas(opFile)
```

Comment:
Code for evaluating filtered replicas is similar in ftsTransfer and in dmTransfer. Might be worth making it a common utility

Reply:
It is the same method that is called...